lib/matplotlib/dates.py

```diff
@@ -738,6 +738,233 @@ def __call__(self, x, pos=0):
+            formatter = mdates.ConciseDateFormatter(locator)
+            ax.xaxis.set_major_locator(locator)
+            ax.xaxis.set_major_formatter(formatter)
+
+            ax.plot(dates, y)
+            ax.set_xlim(lims[nn])
+        axs[0].set_title('Concise Date Formatter')
+
+    """
+
+    def __init__(self, locator, tz=None, formats=None, offsetformats=None,
+            zeroformats=None, show_offset=True):
+        """
+        Autoformat the date labels.  The default format is used to form an
+        initial string, and then redundant elements are removed.
+        """
+        self._locator = locator
+        self._tz = tz
+        self._oldticks = np.array([])
+        self._oldlabels = None
+        self.defaultfmt = '%Y'
+        # there are 6 levels with each level getting a specific format
+        # 0: mostly years,  1: months,  2: days,
+        # 3: hours, 4: minutes, 5: seconds
+        if formats:
+            if len(formats) < 6:
+                raise ValueError('formats argument must be a list of '
+                                 '6 format strings (or None)')
+            self.formats = formats
+        else:
+            self.formats = ['%Y',  # ticks are mostly years
+                    '%b',          # ticks are mostly months
+                    '%d',          # ticks are mostly days
+                    '%H:%M',       # hrs
+                    '%H:%M',       # min
+                    '%S.%f',       # secs
+                    ]
+        # fmt for zeros ticks at this level.  These are
+        # ticks that should be labeled w/ info the level above.
+        # like 1 Jan can just be labled "Jan".  02:02:00 can
+        # just be labeled 02:02.
+        if zeroformats:
+            if len(formats) < 6:
+                raise ValueError('zeroformats argument must be a list of '
+                                 '6 format strings (or None)')
+            self.zeroformats = zeroformats
+        elif formats:
+            # use the users formats for the zero tick formats
+            self.zeroformats = [''] + self.formats[:-1]
+        else:
+            # make the defaults a bit nicer:
+            self.zeroformats = [''] + self.formats[:-1]
+            self.zeroformats[3] = '%b-%d'
+
+        if offsetformats:
+            if len(offsetformats) < 6:
+                raise ValueError('offsetfmts argument must be a list of '
+                                 '6 format strings (or None)')
+            self.offsetformats = offsetformats
+        else:
+            self.offsetformats = ['',
+                    '%Y',
+                    '%Y-%b',
+                    '%Y-%b-%d',
+                    '%Y-%b-%d',
+                    '%Y-%b-%d %H:%M']
+        self.offset_string = 'Dates!'
+        self._formatter = DateFormatter(self.defaultfmt, self._tz)
+        self.show_offset = show_offset
+
+    def __call__(self, x, pos=None):
+        locator_unit_scale = float(self._locator._get_unit())
```

Comment:
I don't know if you are planning to incoorporate the same automatic locator setting as in #12470 here? But even if you don't, it can hit you if for whatever reason you create a 

    ax.xaxis.set_major_formatter(mdates.ConsiseDateFormatter(mticker.MultipleLocator(5)))

or similar.

Reply:
Eh, but you can't do that with `AutoDateFormatter` either, and if you fix the `_get_unit`, then you get caught on the next line: `dmin, dmax = self._locator.viewlim_to_dt()`  I think an argument can be made that we should check for these functions at init time when the locator is passed, but I don't think its reasonable to expect to pass arbitrary Locators to a DateFormatter and expect them to work.

Reply:
I've added a check to AutoDateFormatter and ConciseDateFormatter that error out if a Locator doesn't have `_get_unit`.  OTOH, if you think that other locators can be made date-aware, I'm for making that work.

Reply:
You would be able to do it with `AutoDateFormatter` if #12470 is accepted. In that case `._locator` is only used once to find out the scale to use. If that fails, a default is used.

Reply:
I don't see any change to `AutoDateFormatter` in this PR. 

Anyways, since the ConsiseDateFormatter is newly introduced, one can sure decide to only let it work with locators that have a `_get_unit()`. Clearly, it's better if other locators worked as well. I mean, a `MultipleLocator(5)` isn't totally stupid, if you don't care about the starting day. Or an `MaxNLocator` if you just want to make sure to have *some* non-overlapping labels.

Reply:
OK, this now works w/o throwing an error.  

```python
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker
import matplotlib.dates as mdates
import datetime

fig, ax = plt.subplots()
ax.plot([datetime.datetime(2006,1,1), datetime.datetime(2006,2,1)], [1, 2])
locator = mticker.MaxNLocator(5)
ax.xaxis.set_major_locator(locator)
ax.xaxis.set_major_formatter(mdates.ConciseDateFormatter(locator))
plt.show()

```